### ISSUE TITLE
updating packages in /usr/lib/opencpu/library

### ISSUE THREAD

**Issue by erikriverson** [jeroenooms/opencpu#113]
With the Ubuntu OpenCPU Cloud Server, are there any problems you see with admins keeping the packages in /usr/lib/opencpu/library up to date? 

When a function is called from the OpenCPU Github API, this is the value of .libPaths()

[
	"/tmp/ocpu-www-data/github_library/[[path_to_package]] 
	"/usr/lib/opencpu/library",
	"/usr/local/lib/R/site-library",
	"/usr/lib/R/site-library",
	"/usr/lib/R/library"
]

Consider a package accessed through the OpenCPU Github API (Package A) that imports a package that happens to be contained in /usr/lib/opencpu/library (Package B), and Package B's version is too old to successfully install Package A. 

Using install_github from the R console on the OpenCPU server machine works, since "/usr/lib/opencpu/library" is not in .libPaths() at that point. But then we lose the CI features with OpenCPU/Github webhook integration.

**Comment by followtheart** [comment]
In my case,no package under directory 
> /usr/lib/opencpu/library 
could be loaded in R shell at all

**Comment by alexdeng** [comment]
I run into this issue recently. Why is /usr/lib/opencpu/library needed? .libPath() for the root does not include /usr/lib/opencpu/library  and when updating packages under root it won't see packages in /opencpu/library. This caused a lot of issues for me. For example after updated ggvis, the new version depends on jsonlite >=0.9.11 but the version in /opencpu/library is 0.9.10.

**Comment by alexdeng** [comment]
One way to get around is when updating a package in /usr/lib/opencpu/libary, you need to specify update.packages(lib.loc = c("/usr/lib/opencpu/libary"))

**Comment by jeroenooms** [comment]
The `/usr/lib/opencpu/library` directory contains R packages included with `opencpu-server` that are needed for the opencpu system itself to function. It is deliberately a separate frozen library to ensure that the system is stable and will not be affected by other packages installed or updated on the server.

The only way you should be updating this library is by updating the `opencpu-server` ubuntu package:

    sudo apt-get update
    sudo apt-get upgrade

The latest version of `opencpu-server` will contain the most recent versions of all R packages (including jsonlite 0.9.14).

**Comment by jeroenooms** [comment]
BTW you can see the versions of the latest version of the opencpu library here: https://github.com/jeroenooms/opencpu-server/tree/master/opencpu-lib

**Comment by redmode** [comment]
I got the same issue. It seems that `openCPU` uses older `jsonlite` than my package.

openCPU reports

```
"jsonlite" version 0.9.14 cannot be unloaded.

In call:
library(pkg, character.only = TRUE, logical.return = TRUE, lib.loc = lib.loc)
```

Local `jsonlite` has version 0.9.16

**Comment by jeroenooms** [comment]
Yes that could be an issue. I'll release a new version of opencpu soon with the latest version of jsonlite. But does your package really need version 0.9.16? It will probably work fine with earlier versions of jsonlite as well?

**Comment by redmode** [comment]
Yes, surely I can live with '0.9.14'

I overame it with `devtools::install_version(package = "jsonlite", version = "0.9.14")`

**Comment by jeroenooms** [comment]
I pushed an update for opencpu to launchpad. Your version of jsonlite should be updated the next time you run:

```bash
sudo apt-get update
sudo apt-get upgrade
```

**Comment by redmode** [comment]
Great, thank you!

**Comment by rubenarslan** [comment]
I was just bitten by this "cannot be unloaded problem" with knitr 1.10. Using `devtools::install_version(package = "knitr", version = "1.9", )` I could fix it, thanks for the tip @redmode .

But a more principled solution would be nice. The way I use openCPU this behaviour easily wreaks havoc (i.e. I just updated packages and suddenly I can't use knitr anymore).

**Comment by jeroenooms** [comment]
I have started to push new version to `opencpu-1.5` based on R 3.2.0-patched:

```
sudo add-apt-repository
sudo apt-get update
sudo apt-get dist-upgrade
```

I am reluctant to push a major update into `opencpu-1.4` that might break existing deployments.

**Comment by rubenarslan** [comment]
Don't you mean?

```bash
sudo add-apt-repository ppa:opencpu/opencpu-1.5
sudo apt-get update
sudo apt-get upgrade
```

I think dist-upgrade would upgrade to Utopic Unicorn, that's not desired, is it?

I'm setting up a new server at the moment, so I'll try this out and report back if I run into trouble.

**Comment by rubenarslan** [comment]
Ok, so I had to remove the old knitr version that I just installed (see my first comment here), because now openCPU was trying to unload it's version to load the old version and failing.

Is it good practice not to (re?)install any packages in this list `names(installed.packages(lib.loc="/usr/lib/opencpu/library")[,1])`? 
I might remove all packages in this list to prevent future problems when I simply, stupidly call `update.packages(ask=F)`.

**Comment by jeroenooms** [comment]
Yes, `/usr/lib/opencpu/library` is reserved for the system library, i.e. the packages that opencpu itself needs to function. Usually the only way you want to update this is using `apt-get upgrade`.

**Comment by jeroenooms** [comment]
@rubenarslan The `apt-get dist-upgrade` is terribly named, it does not upgrade your release (`do-release-upgrade` is for that). It is actually very similar to `apt-get upgrade` but also upgrades packages if their dependencies have changed.

**Comment by rubenarslan** [comment]
Oh, ok. I thought I'd seen something about utopic when first running `dist-upgrade` and then cancelled it. But seems to work fine.

Regarding `/usr/lib/opencpu/library` – I wasn't (planning on) updating those packages by hand. 
What I've done is that I've installed knitr, pander, httr and some other packages that are part of openCPU myself (into site-library). And in the case of knitr, the version mismatch caused this unloading error (whenever `library(knitr)`) was called via openCPU.

So, now I was thinking: it probably does not make sense to install any packages into site-library that are already part of openCPU. 
However, if one also uses RStudio Server, then it again makes sense to install e.g. knitr into site-library, because RStudio Server looks there, but not in `/usr/lib/opencpu/library`. 

What I'm looking to prevent is shooting myself in the foot by inadvertently causing such a version mismatch (even if this failed more gracefully which I'm not sure is possible, this would be a waste of time, I suppose).
Maybe you could also share the procedure you use for installing packages on public.opencpu.org.

**Comment by Granjow** [comment]
What is the correct way of installing an R package which depends on packages already present in `/usr/lib/opencpu/library`?

We use jsonlite, knitr, and base64enc, which are already used by OpenCPU, and xtable and reshape, which are not. Those obviously need to be installed – to `/usr/local/lib/R/site-library`, I assume. jsonlite, though, already exists, and when trying to install our package with

    sudo R CMD INSTALL our-package.tar.gz --library=/usr/local/lib/R/site-library

then it complains about missing jsonlite etc. since `/usr/lib/opencpu/library` is not in R’s library paths.

Reason I’m asking is that calling package functions from OpenCPU fails:

    Package 'jsonlite' version 0.9.17 cannot be unloaded
    In call:
    library(pkg, character.only = TRUE, logical.return = TRUE, lib.loc = lib.loc,
        quietly = quietly)

(Why does it have to be *un*loaded anyway?)

Any hints appreciated!

**Comment by jeroenooms** [comment]
@Granjow which version of opencpu are you running?

**Comment by Granjow** [comment]
@jeroenooms It is an Ubuntu 14.04 on 1.5.1-trusty5. There is still R installed (r-base 3.2.2-1trusty0), and we first installed all packages simply via R and then installed opencpu.

**Comment by jeroenooms** [comment]
I don't know why it is trying to unload jsonlite, that should not happen. The best way to install packages in the `site-library` is just installing your dependencies there as well. Try:

```
sudo -i
R
```
And then the usual:

```r
install.packages("jsonlite")
```

**Comment by Granjow** [comment]
Okay; I have done

    uninstall.packages('jsonlite') 

and all other packages that were installed, than re-installed them, and now the error about jsonlite is gone. This does not really tell me anything about the reason – but it works :)

**Comment by Granjow** [comment]
Another update. I tried reproducing the results on a fresh server, and I found that this did not work anymore.

I am now installing the dependencies via an R script and sudo, which installs them to `/usr/local/lib/R/site-library`. jsonlite has to be installed as well because otherwise our-package does not want to be installed. Then I have to delete `/usr/local/lib/R/site-library/jsonlite` again.

But probably it would be better to somehow include `/lib/opencpu/library/` in the library paths when starting R and installing the package, so it does not need to be installed a second time?

**Comment by rubenarslan** [comment]
strongly agree with @Granjow, if that would solve it, it should be in the libpath.
this keeps biting me when I update a package via sudo -i R, then I have to manually deleted doubled dependencies..

**Comment by jeroenooms** [comment]
You know that you can use e.g. `.libPaths("/usr/lib/opencpu/library/")` before installing a package if you want to add a certain library?

**Comment by Granjow** [comment]
I am not very familiar with R in those regards. Can I also do that when installing from a local .tar.gz package with ``R CMD INSTALL``? Because those commands are somehow outside the R environment.

**Comment by jeroenooms** [comment]
Starting opencpu release 1.5.3 the packages from the opencpu library will be symlinked into `/usr/lib/R/library` and therefore also visible in a regular R session.

**Comment by Granjow** [comment]
Thank you! Much better now!

**Comment by rubenarslan** [comment]
some code for the next time this bites me on my old server

```
ocpubasics = names(installed.packages(lib.loc!="/usr/lib/opencpu/library")[,1])
userpkgs = names(installed.packages(lib.loc="/usr/local/lib/R/site-library")[,1])

remove.packages(userpkgs[ userpkgs %in% ocpubasics])
```

**Comment by rubenarslan** [comment]
@jeroen Isn't there a better way to fix this? This is still plaguing me with 2.0.6 and it's a really annoying problem to have because you can accidentally cause this whenever you add a package, without initially noticing an error.
